docs/service-bundle.md

```diff
@@ -0,0 +1,225 @@
+    "cluster": "kubernetes",
+    "namespace": "default"
+}
+```
+
+* _apb_provision_creds: A set of credentials and related data created during a
+  provision action for this service instance. Details are below in the Output
+  section.
+
+### Environment Variables
+
+The following environment variables are set by the broker:
+
+* __POD_NAMESPACE__: the namespace in which the service bundle is being run
+* __POD_NAME__: the name of the pod 
+
+## Output
+
+### stdout / stderr
+
+These are ignored.
+
+### Exit Code
+
+* 0 - success
+* 1 - error
+* 8 - action not implemented/supported
+
+### Encoded Credentials
+
+Binding credentials may be returned during a provision or bind operation by
+creating a named secret (POD_NAME) in the sandbox namespace (POD_NAMESPACE).
+The secret should have a single key called “fields” whose value is a
+base64-encoded JSON document.
+
+Ansible Playbook Bundles use the “asb_encode_binding” module to do this.
+
+```
+apiVersion: v1
+data:
+  fields: eyJEQl9OQU1FIjogImZvbyIsICJEQl9QQVNTV09SRCI6ICJsZXRtZWluIiwgIkRCX1RZUEUiOiAicG9zdGdyZXMiLCAiREJfUE9SVCI6ICI1NDMyIiwgIkRCX1VTRVIiOiAiYmFyIiwgIkRCX0hPU1QiOiAicG9zdGdyZXNxbCJ9
+kind: Secret
+metadata:
+  creationTimestamp: 2018-02-26T18:34:50Z
+  name: apb-874ae78d-acfe-4181-b064-124d6e13c592
+  namespace: dh-postgresql-apb-prov-md5xt
+  resourceVersion: "11846"
+  selfLink: /api/v1/namespaces/dh-postgresql-apb-prov-md5xt/secrets/apb-874ae78d-acfe-4181-b064-124d6e13c592
+  uid: ba79e27a-1b23-11e8-aebf-94b96c9ea8ca
+type: Opaque
+```
+
+The value for “fields” is a base64-encoded JSON document whose keys and values
+represent any data the APB wants to return to the broker and ultimately to a
+client that retrieves a binding.
+
+__Provision Credentials__ - If a service bundle saves such a secret during the
+provision operation, the broker saves the passed data as “provision
+credentials”. That data gets passed into all further actions as part of the
+JSON document under the key ``_apb_provision_creds``, as a sub-document. As such,
+it can be useful for making root-level credentials created during provision
+available to other operations that need to use them. If a request for non-async
+binding is made and the bundle does not run, any “provision credentials” that
+were saved will be returned to the client as binding credentials.
+
+__Binding Credentials__ - If a service bundle saves such a secret during the bind
+operation, the broker saves the passed data as “binding credentials”. These
+credentials get returned to the client that initiated that specific binding
+operation.
+
+## Runtime
+
+The container will be run in a pod by itself with ``restartPolicy: Never``. The
+pod will run in a sandbox namespace created by the broker for that specific
+operation only. The namespace will be deleted at some point after the pod
+finishes.
+
+APB is run with a service account granted access to the ``POD_NAMESPACE`` and
```

Comment:
Should we also discuss the how the permissions are given to the APB? 

Eventually, APB authors are going to need to know the permissions they are running at right?

Reply:
Yeah I agree with Shawn. Worth calling out in this document.

Reply:
Good call. Would you mind supplying that info? Or pointing me to the right place where I can find it?

Reply:
Permission for the service account comes from the `openshift.sandbox_role` from the config. The default for the template is `edit` but I think that most things deploy with this set as `admin` IIRC. here is a link to the roles https://kubernetes.io/docs/admin/authorization/rbac/#user-facing-roles

Reply:
I'm reading up on all of this, including this page: https://github.com/openshift/ansible-service-broker/blob/master/docs/administration.md

But I think I have too many questions in my mind about how things fit together to write this right now. Would one of you mind writing a couple of sentences about this?